lib/spack/spack/stage.py

```diff
@@ -725,6 +743,202 @@ def _add_to_root_stage(self):
+
+    def _remove_context_link(self):
+        self.root_stage_context.unlink()
+
+    def _rebuild_remote_stage(self):
+        full_subdir = Path(CMakeBuildStage.dispatch[self._hash])
+        if not full_subdir.exists():
+            try:
+                full_subdir.mkdir()
+            except FileExistsError:
+                full_subdir = self._setup_remote_build_stage()
+        return full_subdir
+
+    def _setup_remote_build_stage(self):
+        # try to create root if it doesn't exist
+        self._remote_root.mkdir(parents=True, exist_ok=True)
+        sub_dir = self._compute_next_open_subdir()
+        full_subdir = self._remote_root / sub_dir
+        attempts = 0
+        while True and attempts < 200:
+            try:
+                full_subdir.mkdir()
+                break
+            except FileExistsError:
+                # another process must have created the same directory as us
+                # try again
+                attempts += 1
+
+        CMakeBuildStage.dispatch[self._hash] = str(full_subdir)
+        return full_subdir
+
+    def _compute_next_open_subdir(self, last=None):
+        def inc(c):
+            curr = ord(c[-1])
+            over = curr // 122
+            if over:
+                return (inc(c[:-1]) if len(c[:-1]) else "a") + "a"
+            return c[:-1] + chr(curr + 1)
+
+        sort_key = lambda x: (len(x.name), x.name)
+        current_ext_stages = list(self._remote_root.iterdir())
+        if not current_ext_stages:
+            # no currently extant external stages, start enumerating with 'a'
+            return "a"
+        last = sorted(current_ext_stages, key=sort_key)[-1]
+        return inc(last.name)
+
+    def _return_to_stage(self):
+        """Copy external build tree back to stage in normal CMake build dir location"""
+
+        install_tree(str(self._remote_stage), str(self.root_stage_context))
+
+    def _teardown_remote_stage(self):
+        """Destroy external build tree if not keep-stage
+        Otherwise this is kept as usual"""
+
+        def teardown(pth: Path):
+            for sub_item in pth.iterdir():
+                if sub_item.is_dir():
+                    teardown(sub_item)
+                else:
+                    sub_item.unlink()
+            pth.rmdir()
+
+        if self._remote_stage.exists():
+            teardown(self._remote_stage)
+        self._remote_stage = None
+
+    def _reclaim_remote_stage(self):
+        # another Spack process or build has taken this directory
+        # the cmake build will not work from a different dir
+        # so wait until we can take it - try five times
+        # waiting a little longer each time.
+        # This will cause a hang but this should only be called if we're trying
+        # to rebuild a pre-existing stage, so we need to get the previous
+        # build dir or CMake will error
+        ii = 0
+        while self._remote_stage.exists() and ii < 5:
```

Comment:
Depending on where the spack process is terminated (e.g. CTRL-C), `_remote_stage` may be `None`.

Also, how can it occur that the directory would still exist at this point? Is this because you could do something like:

```
# on windows
shutil.rmtree(a_directory)
mkdirp(a_directory) # this fails because the FS on windows hasn't actually deleted the directory
```

?